pkg/api/index.go

```diff
@@ -109,11 +112,23 @@ func (hs *HTTPServer) getAppLinks(c *models.ReqContext) ([]*dtos.NavLink, error)
 			}
 
 			if include.Type == "dashboard" && include.AddToNav {
-				link := &dtos.NavLink{
-					Url:  hs.Cfg.AppSubURL + "/dashboard/db/" + include.Slug,
-					Text: include.Name,
+				if dashboards == nil {
+					if dashboards, err = hs.PluginManager.GetPluginDashboards(c.OrgId, plugin.Id); err != nil {
```

Comment:
this is not a great option, as this fires SQL queries. Doing this inner loop for every plugin & include. Can we not just add UID or url to the include definition, then we do not need to do any lookup here

Reply:
Adding UID or url to the include definition is the way forward for the future, but we still need to support the legacy way of using slug to be backward compatible and solve the existing bug with app plugin dashboard links doesn't work from the side menu. Tricky part is that UID's are not available until the dashboard has been imported/saved to the database, this normally happens after an app plugin has been enabled.

Maybe we can move this code somewhere else, like when loading plugins? My initial idea to this was to short-term provide a `/api/plugins/<plugin>/dashboard/<slug>` that would redirect to the proper dashboard URL. But not optimal either.

Reply:
In practice, the SQL query uses indices and according to the tracing information it doesn't spend much time on a query than what it spends on passing the marshalled response back through all the web framework reflection layers.

I do agree that this is suboptimal, and would have performance impact which linearly increases from the number of plugins that contain dashboards to be included to the menu. In practice, though, with a number of the plugins around ~20 the performance difference is barely noticed by the tracer.

Introducing a UID to the include definition and encouraging plugin developers to use it would remove the performance aspect completely. Introducing another API or bringing back the slug API would leave us with yet another API, destined to be deprecated. To my mind, lazy loading, caching - anything that keeps the API free of slugs - is a better compromise long-term.

Reply:
Also, an issue I found (https://github.com/grafana/grafana/issues/36310) makes me think that most users probably don't have too many app plugins with dashboards in the left nav menu, so performance wouldn't be impacted that much for them.

Reply:
>  but we still need to support the legacy way of using slug to be backward compatible and solve the existing bug  

But there is no check here if the include already contains UID it always performs this query, not just when it is needed. 

This looks like it can be better done on startup when then plugin JSON file is loaded, then it can lookup any UID  and add that the include definition. 

> Trcky's part is that UID's are not available until the dashboard has been imported/saved to the database, this normally happens after an app plugin has been enabled.

That is indeed tricky. Maybe we can require a UID in app  dashboards for these links to work? 

> In practice, the SQL query uses indices and according to the tracing information it doesn't spend much time on a query than what it spends on passing the marshalled response back through all the web framework reflection layers.

Might be true on a local sqlite lite db but might not be true in Grafana cloud. Might move to other storage solutions other than SQL and doing queries in a loop like this could be a bit more expensive. But that should not be relevant here (premature optimization). But this just feels wrong :)  But I see now that it only calls this once per plugin not once per plugin & include ?